File: 704-binary-search/704-binary-search.rs
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        use std::cmp::Ordering;        
        let mut left: i32 = 0;
        let mut right: i32 = nums.len() as i32 - 1;
        while left <= right {
            let mid: i32 = (left + right) / 2;
            match nums[mid as usize].cmp(&target) {
                Ordering::Equal => return mid,
                Ordering::Less => left = mid + 1,
                Ordering::Greater => right = mid - 1
            }
        }
        -1
    }
}
/*
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        use std::cmp::Ordering;
        let (mut low, mut high) = (0, nums.len());
        while low < high {
            let mid = low + (high - low) / 2;
            match nums[mid].cmp(&target) {
                Ordering::Equal => return mid as i32,
                Ordering::Less => low = mid + 1,
                Ordering::Greater => high = mid,
            }
        }
        -1
    }
}
*/

File: 704-binary-search/704-binary-search.cpp
class Solution {
    bool condition(vector<int> &nums, int idx, int tgt) {
        if (nums[idx] >= tgt)
            return true;
        else
            return false;
    }
public:
    int search(vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1, mid = 0;
        
        while (left < right) {
            mid = left + (right - left) / 2;
            if (condition(nums, mid, target)) {
                right = mid;
            } else {
                left = mid + 1;    
            }
        }
        if (nums[left] == target)
            return left;
        else 
            return -1;
    }
};